src/lib/offline-api/validator/resource-links.ts

```diff
@@ -0,0 +1,54 @@
+import { ApiHook } from '../'
+import { ContentTypePayloadValidator } from './content-type'
+import { InvalidActionError, PayloadValidationError } from '../../interfaces/errors'
+import { isResourceLink, MAX_RESOURCE_LINKS } from '../../utils/resource-links'
+import errorMessages from './errors'
+
+export default class ResourceLinksValidator implements ContentTypePayloadValidator {
+  public hooks = [ApiHook.SaveContentType]
+
+  public validate({ contentType }): (InvalidActionError | PayloadValidationError)[] {
+    const errors: (InvalidActionError | PayloadValidationError)[] = []
+
+    const fields = contentType.fields.toRaw()
+    const resourceLinkCount = fields.filter(isResourceLink).length
```

Comment:
You can use the filtered array below then you would not need `isResourceLink(field) &&` in your if conditions.

```
const resourceLinks = fields.filter(isResourceLink)
if (resourceLinks.length > MAX_RESOURCE_LINKS) {
   ...
}

....

for (const resourceLink of resourceLinks) {
      if (!resourceLink.allowedResources) {
         ....
```

Reply:
Then I’d need two filtered arrays and iterate over both which doesn’t seem like a better solution to what I’m doing now.

Reply:
I don't understand why you would need 2 filtered arrays. It's just replacing the `fields` in the `for of` with `resourceLinks`.  It would be like the snippet below. Am I missing something?

```
    const fields = contentType.fields.toRaw()
    const resourceLinks = fields.filter(isResourceLink)

    if (resourceLinks.length > MAX_RESOURCE_LINKS) {
      errors.push({
        type: 'InvalidPayload',
        message: errorMessages.contentType.TOO_MANY_RESOURCE_LINKS(
          contentType.id,
          MAX_RESOURCE_LINKS
        )
      })
    }

    for (const field of resourceLinks) {
      if (!field.allowedResources) {
        errors.push({
          type: 'InvalidPayload',
          message: errorMessages.field.REQUIRED_DEPENDENT_PROPERTY(
            'allowedResources',
            field.id,
            'type',
            'ResourceLink'
          )
        })
      }

      if (field.allowedResources) {
        errors.push({
          type: 'InvalidPayload',
          message: errorMessages.field.FORBIDDEN_DEPENDENT_PROPERTY(
            'allowedResources',
            field.id,
            'type',
            'ResourceLink'
          )
        })
      }
    }

    ....
}

Reply:
Your code creates a `FORBIDDEN_DEPENDENT_PROPERTY` error for resource links which doesn’t make sense. Check the original code: one `isResourceLink` has a `!` in front of it.

Reply:
Yes you're right. I read it wrong.